packages/react/src/__tests__/state.test.tsx

```diff
@@ -0,0 +1,273 @@
+})
+
+it('should set state with custom value', async () => {
+  const value = state('foo')
+  const dataSource: DataSource = {
+    get: () => ({
+      subscribe: () => jest.fn(),
+      get() {
+        return [
+          {
+            $: 'component',
+            key: 'header',
+            component: 'View',
+            children: value,
+          },
+          {
+            $: 'component',
+            key: 'button',
+            component: 'View',
+            props: {
+              ['data-testid']: 'button',
+              onClick: event(setStateAction(value, 'custom-string-value')),
+            },
+          },
+        ]
+      },
+    }),
+    sendEvent: jest.fn(),
+  }
+  const component = render(<Template components={BUILT_IN_COMPONENTS} dataSource={dataSource} />)
+  await act(async () => {
+    fireEvent.click(component.getByTestId('button'))
+  })
+  expect(component.asFragment()).toMatchInlineSnapshot(`
+    <DocumentFragment>
+      <div>
+        custom-string-value
+      </div>
+      <div
+        data-testid="button"
+      />
+    </DocumentFragment>
+  `)
+})
+
+it('should toggle state', async () => {
+  const isDisabled = state(false)
+  const dataSource: DataSource = {
+    get: () => ({
+      subscribe: () => jest.fn(),
+      get() {
+        return [
+          {
+            $: 'component',
+            key: 'header',
+            component: 'View',
+            props: {
+              disabled: isDisabled,
+            },
+          },
+          {
+            $: 'component',
+            key: 'button',
+            component: 'View',
+            props: {
+              ['data-testid']: 'button',
+              onClick: event(setStateAction(isDisabled, toggle)),
+            },
+          },
+        ]
+      },
+    }),
+    sendEvent: jest.fn(),
+  }
+  const component = render(<Template components={BUILT_IN_COMPONENTS} dataSource={dataSource} />)
+  expect(component.asFragment()).toMatchInlineSnapshot(`
+    <DocumentFragment>
+      <div />
+      <div
+        data-testid="button"
+      />
+    </DocumentFragment>
+  `)
+  await act(async () => {
+    fireEvent.click(component.getByTestId('button'))
+  })
+  expect(component.asFragment()).toMatchInlineSnapshot(`
+    <DocumentFragment>
+      <div
+        disabled=""
+      />
+      <div
+        data-testid="button"
+      />
+    </DocumentFragment>
+  `)
+})
+
+it('should increment state', async () => {
+  const counter = state(1)
+  const dataSource: DataSource = {
+    get: () => ({
+      subscribe: () => jest.fn(),
+      get() {
+        return [
+          {
+            $: 'component',
+            key: 'header',
+            component: 'View',
+            children: counter,
+          },
+          {
+            $: 'component',
+            key: 'button',
+            component: 'View',
+            props: {
+              ['data-testid']: 'button',
+              onClick: event(setStateAction(counter, increment(2))),
```

Comment:
nice!

Reply:
thanks! it's also type-safe, so you can't pass `toggle` to a string state or `increment`. Only to what is compatible. If you ignore it, you will receive some strange (most likely) outcomes!